### ISSUE TITLE
Enable prisma migrate dev to ignore specified DB tables

### ISSUE THREAD

**Issue by janconc** [prisma/prisma#8864]
## Problem

I am developing an application that uses Prisma to facilitate DB operations with DB/tables and I am using Prisma migrations to create or alter those tables (this part works OK).

But the application also uses dynamic tables (their names and structure are not yet defined) which are not defined by Prisma schema and will not be created by Prisma migrations. For local development and testing I have created one such table in the database (SQL server).

When running `prisma migrate` I get an error about database drift due to the existing table that prisma doesn't have a clue about but which I need.

## Suggested solution

Add a command line parameter for `prisma migrate dev` to ignore specified tables.

## Alternatives

At the moment I have created a special DB user which is denied the access to the table I need to ignore. While it works it is not exactly user friendly.

**Comment by nickluger** [comment]
Brief addendum for ppl landing here from Google, and also my future self (landing here from Google):

Of course, you need to make sure that the `prisma` role does **not own** the schema itself, only tables within the schema. Being owner of the schema would still prevent the `prisma` role from accessing tables owned by others. So you would not see db drift warnings initially, but running `prisma migrate` commands would drop the whole schema ([`DROP CASCADE`](https://github.com/prisma/prisma-engines/blob/800fbfbed33df701b0741db1f0a62c4631c43b0a/migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs#L359)), including the tables **not** owned by `prisma`. This is not what you want.

`prisma migrate reset` and reset induced by `prisma migrate dev` (migration files changed) should still work. If Prisma migrate cannot drop the whole schema, it will "cleanup" all tables and other stuff (best effort reset), as can be seen here [in the engine](https://github.com/prisma/prisma-engines/blob/800fbfbed33df701b0741db1f0a62c4631c43b0a/migration-engine/connectors/sql-migration-connector/src/lib.rs#L107). As it works on a table-level (not schema level), access to other user's tables should be restricted, so these are not dropped.

**Comment by janconc** [comment]
Hi, sorry for delay, I have added example of my workaround into the original post.

**Comment by janpio** [comment]
> What I was wondering, without knowing prisma internals, there must be already implemented something similar because **_prisma_migrations** table is ignored (not introspected).

Yes, but that is one hard coded table where we also know that it does not have relations with other tables (which is what leads to a lot of complexity in designing a feature like this).

**Comment by michaelbuller** [comment]
`npx prisma migrate dev --create-only`
Environment variables loaded from .env
Prisma schema loaded from libs\api\core\data-access\src\prisma\schema.prisma
Datasource "db" - SQL Server

Drift detected: Your database schema is not in sync with your migration history.

The following is a summary of the differences between the expected database schema given your migrations files, and the actual schema of the database.

It should be understood as the set of changes to get from the expected schema to the actual schema.

[+] Added tables
  - stage_Providers

[*] Changed the `ProviderSearch` table
  [+] Added column `geography`

? We need to reset the database.
Do you want to continue? All data will be lost. » (y/N)

My table has   geography         String? @ignore  on it, the geography type is not supported by prisma yet... so have to work around it somehow.  How do I tell Prisma not to touch that column and move on?  Also, this is a production database and people need to move data into it via different methods.  Seen here, there is a stage table that was added.  We need a way to ignore that table.  Also, don't you think it is odd that your only option you present to the user when a "New" table is added is to DELETE all of the other data from the database?  Even if this was development, and we have all of the Prisma seed methods in place, deletion all data doesn't seem like the first choice I would go with.

**Comment by dvins** [comment]
We love Prisma and really want to like Prisma Migrate, but In lieu of some basic attributes to be able to specify tables (or views or other elements) to ignore we're considering falling back to Knex (or another mechanism entirely) for all table schema work. There are just too many hoops to jump through when it does not have total control over the entire namespace. Unfortunately, that's not an option for us with disparate microservices and shared query space. We're already having hard enough time with the one Prisma schema file to rule them all.

**Comment by janpio** [comment]
@michaelbuller Please open a separate issue and describe your full problem re `ProviderSearch.geography`, including the current SQL schema and Prisma schema so we can try to understand this and then see if this is intended - or a bug - and help you figure out a way around this.

And yes, we realize this is a use case some people have, hence this feature request issue exists. We collect use cases and input that allows us to prioritize our work.

**Comment by janpio** [comment]
> would like to use prisma with 1-4 existing tables, in an existing database with 50+ tables. There has to be a better way?

Not right now. But this makes a good feature requests for that, so I updated the title of the issues to cover both - ignoring specified tables or only looking at specified tables.

**Comment by juzerzarif** [comment]
@janpio Will this feature go on the roadmap in the near future? I understand that it's not necessarily an easy ask and the suggested workaround here works for now but it would be nice to know if this is something we can expect to land in a future release as a native feature?

**Comment by floelhoeffel** [comment]
@juzerzarif unfortunately we don't know yet. But as long as there is a workaround, we will most likely focus on features without workarounds in the next few months.

https://www.prisma.io/docs/about/prisma/roadmap

**Comment by janpio** [comment]
That is very good to know @juzerzarif - thanks for clarifying.

**Comment by janpio** [comment]
No, this happens in the Rust Engine part of Prisma. Please make sure to open another issue for the underlying feature request.

**Comment by nvitaterna** [comment]
https://github.com/prisma/prisma-engines/blob/15401d37712d3df91d04e8824690855a1306f829/migration-engine/connectors/sql-migration-connector/src/sql_schema_differ/sql_schema_differ_flavour.rs#L147-L149

**Comment by niederschlag** [comment]
Found the spot where a workaround is possible: https://github.com/prisma/prisma-engines/blob/main/migration-engine/connectors/sql-migration-connector/src/sql_schema_differ/differ_database.rs#L42-L43

Thanks, @nvitaterna ! You almost got it right but that wouldn't fix Postgres schema diffing.

The main problem there is the missing `prisma.schema` DML context. The diff compares two databases internally, so there are only raw tables, columns, etc... no `@@ignore`-information at all.

I've patched my `migration-engine`, so a local `.prismaignore` contains tables that are being ignored in migration-diffing. It's still only a workaround, so not sure if I should create a PR? It would certainly be better if this information was inside the `prisma.schema` but no clue how to parse it from the `migration-engine`

**Comment by janpio** [comment]
The PR would be fine for informational purposes for us and users, who might want to copy your approach. We would probably not consider merging it though, as we want and need to _design_ the solution here to indeed be more integrated (e.g. by being via the schema file).

**Comment by chaiwa-berian** [comment]
Yes, this feature would be helpful. I also want to add new tables to an already exisiting database and only want to manage/version the new tables using Prisma without affecting the rest of the tables in the database. My application will  only use the new tables but other exiting applications will continue to use the existing tables and the new tables from my App.

**Comment by theMogget** [comment]
The workaround looks great, but unfortunately doesn't seem to work in all cases. `prisma migrate dev` is still detecting drift caused by tables that the user `prisma` doesn't own.

Running `prisma db pull`, I get: `The following models were commented out as we could not retrieve columns for them. Please check your privileges.`

Checking by selecting from `pg_catalog.pg_tables` I can see that the owner is indeed `admin`.

I'm on Postgres 14.2, and Prisma 4.1.1.

**Comment by janpio** [comment]
That sounds like the user has _some_ permissions to see the table, but not the columns in there - just by going by the error message you posted.

**Comment by theMogget** [comment]
> That sounds like the user has _some_ permissions to see the table, but not the columns in there - just by going by the error message you posted.

@janpio I believe that's the case too - I tried the same steps on a fresh new DB and was able to replicate the same behaviour. Maybe an update in Postgres? Using TablePlus, I can't see the `admin`-only tables with the `prisma` user. But using DBeaver I can. If I was more familiar with Postgres I'd try to look through the changelog, but unfortunately I've no idea where to start. 😅 

What I'm doing for now is separating by creating new DB schemas altogether.

**Comment by janpio** [comment]
What commands exactly did you use as a workaround for PostgreSQL @theMogget?

**Comment by dvins** [comment]
@janpio Sorry, but is there any progress to be had on this at all, even real consideration of solving this keystone issue in the near future? The inability to tell Prisma Migrate to ignore artifacts it does not own is causing unbelievable friction. 

The hoops we're all jumping through to handle standard database scenarios, be it views, partitioned tables, stored procedures, etc. because Prisma Migrate does not play well with others is far beyond painful. 

The number of issues connected in some way to this, either directly or indirectly so we can build other work arounds for features not yet supported continues to grow too:

- https://github.com/prisma/prisma/issues/678 (2019)
- https://github.com/prisma/prisma/issues/807) (2019)
- https://github.com/prisma/prisma/issues/2278 (2020)
- https://github.com/prisma/prisma/issues/4510 (2020)
- https://github.com/prisma/prisma/issues/8864 (2021)
- https://github.com/prisma/prisma/issues/12231 (2022)
- https://github.com/prisma/prisma/issues/13407 (2022)
- https://github.com/prisma/prisma/issues/15719 (2022)

Database view support is the oldest, most commented issue in the stack. Many of us can get by without Prisma directly supporting it, but we cannot easily overcome the conflicts with Prisma Migrate because it will not be a friendly neighbor.

**Comment by dvins** [comment]
Even the official workaround guidance for view support no longer works either due to this: 
- https://github.com/prisma/prisma/issues/3933 (2022)

**Comment by janpio** [comment]
We are aware that there are many things we could do, but we are only human and have to prioritize some things or will get _nothing_ done at all as we change focus every 2 days and can not get anything ready enough to actually ship. 
You can see our current priorities in the work we are doing (all issues and pull requests are open) and by following our public roadmap at https://pris.ly/roadmap

I fail to see how pulling in other, unrelated issues to this one here is helpful. This issue here is about Prisma and specifically Migrate ignoring some database tables or not looking at some database tables. I do not understand how the other things you linked to are related to that issue (except maybe #13407). 

If you want to have a higher level discussion about our prioritization, please open a discussion and do not comment in existing unrelated issues. (The response will be pretty much the same as my first paragraph here though - we are aware of these issues and doing our best to work on them in an order that makes sense to our users so we create the most possible amount of value.)

**Comment by arieljr10699** [comment]
If it works for someone else, I've been able to only introspect certain tables from my SQL Server DB by creating a user and granting access (select, insert, update, delete and alter) to the tables I'm working with. Then, in the connection string that Prisma uses, I specify the DB, user, and password I'm using. When I do an instrospect it only pulls the tables with the granted access.

**Comment by dzmitry-kankalovich** [comment]
This was a sad ride of trying to work around this issue in Postgres as of Jan 2023. The suggestion by @nickluger seems to be no longer working in the latest Postgres (at least in AWS Aurora variant). As well as `@@ignore` has no effect on `prisma migrate dev`.

I wonder if there are any other possible workarounds for Postgres specifically? Schema split seems to be the only option, but I don't want to go there.

**Comment by blevine** [comment]
Just wanted to leave a suggestion that this mechanism also work with `prisma db pull`. I use Knex migrations instead of Prisma migrations on an existing DB schema that has many tables I'll never access.  So my workflow is:

`create Knex migration -> execute Knex migration against local DB -> prisma db pull against local DB -> prisma generate against local DB.`

So it would be very helpful to have include/ignore functionality in the `prisma db pull` step.

**Comment by danielbuechele** [comment]
Coming here from the discussion in https://github.com/prisma/prisma/discussions/17606

In my case, I have a column with a foreign key to a table which is not managed through Prisma. I'm hiding the other table by moving it into another Postgres schema. But `prisma db pull` complains that my column has a foreign key to a table which is not part of the schema, even though I used `@ignore` on the column with the foreign key.
Would be super helpful if we could change the `db pull/push` operation to also ignore foreign key constraints in `@ignore`d columns.

**Comment by robert-king** [comment]
I call create_hypertable in a migration to create a timescale hypertable. Prisma creates a second migration that drops the timestamp key created on the hypertable.

**Comment by janpio** [comment]
@danielbuechele That sounds like something that would deserve its own issue - how exactly relations / foreign keys are handled here is unspecified, and the additional issue with reasoning and suggestions could be useful for that.

@robert-king That does not sound related to this problem. Please open a discussion for your problem and describe fully what is going on. Thanks.

**Comment by robert-king** [comment]
I guess I could have told prisma to ignore that database, or the key that it dropped. I found another way to specify that key in prisma so it's okay now :)

**Comment by janpio** [comment]
Another related issue: https://github.com/prisma/prisma/issues/4704

**Comment by Jolg42** [comment]
Another use case with some context:

> ## Problem
> 
> I have tables in my database that have nothing to do with prisma. It is super frustrating when prisma wants to drop these tables when I do a db push. These tables btw are dynamically generated and have dynamic table names so db pull is not viable
> ## Suggested solution
> 
> Add a way to ignore certain tables via wild card. In my case, I want to be able to ignore table names like documents_3347074906 and documents_396766771 by specifying somewhere in the schema to ignore table with names documents_*

**Comment by divmgl** [comment]
I'm running into this issue now as copies of a database I'm restoring has tables and changes that are not in the Prisma schema, but I do want to apply migrations on top. Unfortunately Prisma is complaining about deviations and there's no way to let it know that these deviations are expected.

**Comment by robert-king** [comment]
another issue we hit: we have a team using DBT to transform some data, but we want some of the resultant data to be referenced by a prisma foreign key. Obviously prisma would complain it doesn't know about that table and also wouldn't like the FK.

**Comment by wmtrinu** [comment]
**_for those who use introspection..._**

I have an _authentication service_ that interacts with one table alone (`account table`), since we heavily rely on **introspection**, this is the only table that we need. (our organization doesn't use `prisma migrate`)

**Comment by divmgl** [comment]
> I'm running into this issue now as copies of a database I'm restoring has tables and changes that are not in the Prisma schema, but I do want to apply migrations on top. Unfortunately Prisma is complaining about deviations and there's no way to let it know that these deviations are expected.

I want to add some additional context to this in case the Prisma team is interested in fixing this problem.

I made changes locally to our Prisma schema, but Ihad a local copy of a production database. Because our production database has additional tables and columns that are not present in the Prisma schema (since we migrated from a TypeORM installation and I guess we didn't go a good job of pruning and removing old tables and columns), running `prisma migrate dev` **always** results in having to reset the database in order to create a migration.

Therefore, my workflow is:

1. Make a backup of the database.
2. Run `prisma migrate dev` which resets the database and creates a migration.
3. Drop the database and recreate it.
4. Restore the backup of the database.
5. Then run `prisma migrate deploy` which forces the new migration to apply.

This is all super cumbersome. It'd be nice if Prisma somehow kept track of the resources in the `prisma.schema` file and tracked how they changed over time (which requires an architectural change), **OR** (and this is key to this issue I think) **give us a way to ignore drifts in the database, including tables that are not included in `schema.prisma`**. As mentioned in this issue, using `@@ignore` does not solve the migration issue.

Hope this helps.

**Comment by divmgl** [comment]
Still waiting on the ability to ignore specific tables from a Prisma schema. We're now running into this limitation with two Prisma clients for different schemas. We need to be able to add a foreign key relationship to the other schema, and we want to be able to have it in our new `schema.prisma` file, but if we include it then the migration suite will pick it up every time. 

We can edit the migrations to not include the table changes, but it's annoying.

**Comment by dmnro123** [comment]
is there a workaround for this for now?

**Comment by algoflows** [comment]
Wild this hasn't been solved yet, stuck with cockroach which generates an enum for multiple clusters which prevents me from creating migrations which prevents then managing my cms using prisma. Mental.

**Comment by luixal** [comment]
When doing a _db push_ agaisnt a database which contains tables that are not defined in Prisma schema, Prisma warns you about:

`There might be data loss when applying the changes`

but if you create empty models named after those tables you want to ignore and just add the `@@ignore` , that table disapears from the warning list.

Does that mean Prisma is not doing anything at all with that table? No data loss, etc.?

Thanks!

## Detailed Example

Running `db push` warns you this:

```
There might be data loss when applying the changes:

  • You are about to drop the `table1` table, which is not empty (48 rows).
```

and you add this in the Prisma schema:

```
model table1 {
  @@ignore
}
```

you run `db push` again and no warning message is shown...

**Comment by neighbaa** [comment]
I was running up against this and found a solution that works for me using a basic node script. In my case, I can compromise on running this instead of db push and migrate dev. Not a great compromise, but it works, and perhaps others will find this a useful workaround or be able to tweak it for their own needs.

Caveats:
- If you intend to use this, read the code carefully before executing it as it strips PostgreSQL commands (which may not include all commands) and could lead to data loss.
- written for postgres and it will likely need to be adjusted for other db languages.
- It uses a combination of prisma migrate diff, from the schema db source to the schema, and prisma migrate deploy, so if that doesn't work for you, or you need migrate dev, etc., this won't work.
- this script assumes pnpm, but can be easily switched.
- this script assumes fairly conventional directory structure for prisma.
- if you get stuck on a migration, you'd have to manually delete the migration files and relevant db row.

Create a custom-migrate.js script and a json file to list the tables to ignore (I went with .ignore-prisma-tables.json):

#### **`custom-migrate.js`**
``` javascript annotate
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const tablesToIgnore = require('./.ignore-prisma-tables.json');

function executeCommand(command) {
  try {
    console.log(`Executing: ${command}`);
    execSync(command, { stdio: 'inherit' });
  } catch (error) {
    console.error(`Error executing command: ${command}`);
    throw error;
  }
}

function removeTableReferencesFromMigration(filePath, tablesToIgnore) {
  let content = fs.readFileSync(filePath, 'utf8');
  
  tablesToIgnore.forEach(table => {
    const regexDropTable = new RegExp(`-- DropTable\\s+DROP TABLE (IF EXISTS )?"${table}"[\\s\\S]*?;\\s+`, 'g');
    content = content.replace(regexDropTable, '');
    const regexCreateTable = new RegExp(`-- CreateTable\\s+CREATE TABLE "${table}"[\\s\\S]*?;\\s+`, 'g');
    content = content.replace(regexCreateTable, '');
    const regexDropIndex = new RegExp(`-- DropIndex\\s+DROP INDEX "${table}.*?[\\s\\S]*?;\\s+`, 'g');
    content = content.replace(regexDropIndex, '');
    const regexCreateIndex = new RegExp(`-- CreateIndex\\s+CREATE (UNIQUE )?INDEX "${table}.*?[\\s\\S]*?;\\s+`, 'g');
    content = content.replace(regexCreateIndex, '');
    const regexAlterTable = new RegExp(`-- AlterTable\\s+ALTER TABLE "${table}"[\\s\\S]*?;\\s+`, 'g');
    content = content.replace(regexAlterTable, '');
  });

  fs.writeFileSync(filePath, content);
}

async function main() {
  try {
    // Step 1: Generate a migration using prisma migrate diff
    console.log('Generating migration...');
    const timestamp = new Date().toISOString().replace(/[:.-]/g, '_');
    const migrationName = `ignore_dynamic_tables_${timestamp}`;
    const migrationPath = path.join(__dirname, 'prisma', 'migrations', migrationName);
    fs.mkdirSync(migrationPath, { recursive: true });
    const migrationFilePath = path.join(migrationPath, 'migration.sql');

    executeCommand(`pnpm prisma migrate diff --from-schema-datasource prisma/schema.prisma --to-schema-datamodel prisma/schema.prisma --script > ${migrationFilePath}`);

    // Step 2: Remove references to ignored tables from the migration file
    console.log('Removing references to ignored tables...');
    removeTableReferencesFromMigration(migrationFilePath, tablesToIgnore);

    // Step 3: Create a migration.toml file
    const migrationToml = `# This is an empty migration.\n\nmigration_name = "${migrationName}"\n`;
    fs.writeFileSync(path.join(migrationPath, 'migration.toml'), migrationToml);

    console.log('Migration file created and updated successfully.');

    // Step 4: Apply the modified migration
    console.log('Applying migration...');
    executeCommand('pnpm prisma migrate deploy');

    // also generate
    executeCommand('pnpm prisma generate');

    console.log('Migration completed successfully.');
  } catch (error) {
    console.error('An error occurred:', error);
    process.exit(1);
  }
}

main();
```

#### **`.ignore-prisma-tables.json`**
```json annotate
[
    "some_dynamically_created_table_to_ignore",
    "some_pattern_to_ignore_[a-zA-Z0-9_\\$]+",
]
```

Then just run it like any other node script
`node ./custom-migrate.js`

Some things I like about this:
- I can use table name patterns instead of hard-coded names.
- I can choose not to include the tables in my prisma schema.
- if I want the tables available on the prisma client I can opt to include them in the schema and prisma generate still includes them. In my case I only wanted to exclude my dynamic tables from the migration, but I liked that I can use the prisma client in a standard way with them + I get all the types.

Again, this is a bit of a hacky, imperfect solution, so be careful using it!

**Comment by johannespn** [comment]
> Wild this hasn't been solved yet

Yeah.. wild indeed.. wanted to use prisma in combination with supabase auth, this now forces me to duplicate user data.. :/

**Comment by vasily-polonsky** [comment]
Is there a workaround that works? I have a table that references a table in another schema maintained by another team. So the schema cannot be described by me as I am not a maintainer of that schema. Creating a user does not help as prisma is an owner of created tables, so I am getting  "Cross schema references are only allowed when the target schema is listed in the schemas property of your datasource"

**Comment by Inambe** [comment]
> > Wild this hasn't been solved yet
> 
> Yeah.. wild indeed.. wanted to use prisma in combination with supabase auth, this now forces me to duplicate user data.. :/

Did you find the solution? I'm stuck here too for exactly this reason. I made a profiles table/schema (for storing `username`) and want to have a relation field with `auth.users` without modifying or overwriting it.

**Comment by johannespn** [comment]
> > > Wild this hasn't been solved yet
> > 
> > 
> > Yeah.. wild indeed.. wanted to use prisma in combination with supabase auth, this now forces me to duplicate user data.. :/
> 
> Did you find the solution? I'm stuck here too for exactly this reason. I made a profiles table/schema (for storing `username`) and want to have a relation field with `auth.users` without modifying or overwriting it.

My solution is creating the profile table on login, with the same userId that is stored in auth.users without a relation. As of my research, Prisma does not natively support this. I would look at [drizzle](https://orm.drizzle.team/) next time.

**Comment by zeroclap** [comment]
> This works for Postgres (>=12), but it should work similarly for other RDBMS:
> 
> In general, you need two database users/roles (synonym in Postgres for our purposes).
> 
> Let's call them `prisma` and `admin`. `admin` has access to all tables and schemas, while `prisma` owns and has access to the tables that should be accessed by Prisma client and managed by `prisma migrate`. The user called `prisma` is used to run `prisma migrate` and also the Prisma client requests.
> 
> This script creates a special DB user `prisma` who owns all tables of the `public` schema, except for a few, which you do not want to manage via Prisma. These are hidden from the `prisma` user, by being owned by a different DB user.
> 
> ```sql
> -- Local development: Needs CREATEDB to be able to create shadow DB when running "prisma migrate dev"
> CREATE USER prisma WITH PASSWORD 'password' CREATEDB;
> -- Production: CREATEDB not needed.
> -- CREATE USER prisma WITH PASSWORD 'super-secure-password';
> 
> -- prisma user can connect to DB via prisma client, CLI etc.
> GRANT CONNECT ON DATABASE mydb TO prisma;
> -- prisma user can inspect & create tables in the schema
> GRANT ALL ON schema public TO prisma;
> 
> -- Assuming, existing tables were already created by another user (admin)
> -- You need to transfer ownership to prisma user,
> -- if these tables should be managed by prisma migrate
> ALTER TABLE table_a OWNER TO prisma;
> ALTER TABLE table_b OWNER TO prisma;
> ALTER TABLE table_c OWNER TO prisma;
> -- ... etc.
> ```
> 
> If the user `prisma` creates new tables via `prisma migrate`, it should get full privileges for them automatically, as he is the owner.
> 
> If you were to add or remove tables dynamically as described by @janconc, and another user creates them (e.g. `admin`), the `prisma`-user should automatically not have access to them, i.e. they should not cause drift issues during `prisma migrate dev`.
> 
> Updated this, as you really need to change **ownership** of **existing** tables to the new `prisma` user, otherwise Prisma's `ALTER TABLE` statements will fail, as described [in the Postgres manual](https://www.postgresql.org/docs/current/sql-altertable.html). It's not enough to give all privileges to the user `prisma`.
> 
> > You must own the table to use ALTER TABLE. The right to drop an object, or to alter its definition in any way, is not treated as a grantable privilege; it is inherent in the owner, and cannot be granted or revoked.
> 
> On the other hand, it's not necessary to revoke access from tables, that are not owned by the `prisma` user, as we never `GRANT` any privileges for them.
> 
> This makes this approach super-slick. _If it's made and controlled by Prisma, it can be seen by Prisma. If somebody else made it, it can't_

Unfortunately it's not working for me on the postgres.  Once user has access to the schema it will be able to check what sort of tables exist (and the structure of tables, probably), even without access to the data itself.

However, I have found a quite generic approach (I guess). Instead of ignoring tables or making limited access user (which still a good practice for the access management), what you can do - you can create a separate schema just for migration generation purposes (`prisma migrate dev`) and once the migration is generated, you can back to your original schema and run `prisma migrate deploy` to apply generated migration

Hope it will be useful for someone

**Comment by johannespn** [comment]
Asking again since I am considering to use prisma for another project. Is there any way to integrate an external table that is not managed by prisma, like `auth.users`, with a table managed by prisma like `public.users`?

**Comment by dubem-design** [comment]
Still needed today

**Comment by ihr-it-projekt** [comment]
Will be great that prisma can ignore tables

**Comment by michaelbuller** [comment]
It does ignore tables. You should have a local dev copy that is empty and has only the prisma shapes in there. Generate the migrations from this and then switch connection string to database you want to update, then ‘prisma migrate deploy’. Sent from my iPhoneOn Apr 30, 2025, at 7:50 AM, Ralf Geisthardt ***@***.***> wrote:﻿
ihr-it-projekt left a comment (prisma/prisma#8864)
Will be great if prisma can ignore tables

**Comment by ihr-it-projekt** [comment]
Okay, maybe I wasn't precise enough. We use directus as a headless cms. These have their own migrations and their own tables. We have created our own content in directus, which generates tables. We now want to ignore all tables that have the prefix directus__ so that we can create migrations that only reflect our changes. Put simply, the directus__ tables should not appear in the schema.

**Comment by nvitaterna** [comment]
> Okay, maybe I wasn't precise enough. We use directus as a headless cms. These have their own migrations and their own tables. We have created our own content in directus, which generates tables. We now want to ignore all tables that have the prefix directus__ so that we can create migrations that only reflect our changes. Put simply, the directus__ tables should not appear in the schema and generated migrations.

This is my use case as well.

**Comment by ihr-it-projekt** [comment]
@nvitaterna Have you solved it, or does you also search for a solution?

**Comment by michaelbuller** [comment]
1. Create a new database that is completely empty. This is typically a
   local Dev machine.
   2. Update the Connection string in your ENV file to point here.
   3. `prisma migrate deploy` (This creates the database that Prisma will
   use to compare against so that it knows the migrations that should apply)
   4. Make a change to the primsa.schema
   5. `prisma migrate dev --create-only` and then name it
   6. `prisma migrate deploy` this will apply the migration to the dev
   database.
   7. update the connection string to production
   8. `prisma migrate deploy` this command is non destructive and will only
   make changes based on the migrations that have not been deployed.

No where in this process do you use `db push` or dev commands. Those are
not for Production workloads.

**Comment by divmgl** [comment]
It's been almost three years at this point. Are there plans to add this basic feature into Prisma? We need to be able to ignore existing tables.

**Comment by raress96** [comment]
Damn, another basic feature that Prisma doesn't have...

At this point I think Prisma is not even ready for production use, at least not in real applications. Maybe for side projects it is fine, but already live production apps use other systems that may integrate with the same database, and Prisma not being able to ignore tables in migrations is very strange...

**Comment by cesarvspr** [comment]
Seems like prisma priority is not listen to community. They seems very focused on their serveless db solution

**Comment by AmineYagoub** [comment]
> Seems like prisma priority is not listen to community. They seems very focused on their serveless db solution

Haha, it's money time.

**Comment by dvins** [comment]
> Seems like prisma priority is not listen to community. They seems very focused on their serveless db solution

Ironically, it's ignored features like this which for years have blocked enterprise adoption because there's no way to incrementally introduce the tech into the stack without disrupting it.

**Comment by adwinang** [comment]
Is there any updates regarding this?

**Comment by marcellov7** [comment]
any news?

**Comment by pkp-pkp** [comment]
This is a solution for migration offset. I'm not sure if it's useful to you when I have time. For your convenience, the following language is in Chinese. Please translate it yourself

我视乎有个可能有用的解决方法，使用不同的 .env环境，创建一个 .env_migrate 环境并指向一个新的数据库，让prisma迁移始终在新的数据库中做操作，这样可以避免丢失数据等一系列问题，然后再将生产的迁移文件运用到当前开发数据库中，这样可以解决动态生成的表，被prisma迁移检测偏移
首先你需要创建一个 
`.env_migrate` 文件，写入 `DATABASE_URL="mysql://root:abc123@localhost:3306/adplatform_migrate"`
创建以下脚本，然后去执行他

`#!/usr/bin/env node

/**
 * 干净迁移脚本
 * 
 * 解决动态表导致Prisma迁移生成失败的问题
 * 
 * 工作流程：
 * 1. 使用 .env_migrate 连接到干净的数据库生成迁移文件
 * 2. 使用正常的 .env 连接到实际数据库部署迁移
 * 
 * 使用方法：
 * npm run prisma:clean-migrate [migration-name]
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

class CleanMigrate {
  constructor() {
    this.migrationName = process.argv[2] || 'auto_migration';
    this.migrateDbUrl = this.getMigrateDbUrl();
    this.normalDbUrl = this.getNormalDbUrl();
  }

  /**
   * 从 .env_migrate 获取迁移数据库URL
   */
  getMigrateDbUrl() {
    const envMigratePath = path.join(process.cwd(), '.env_migrate');
    if (!fs.existsSync(envMigratePath)) {
      throw new Error('❌ .env_migrate 文件不存在，请先创建迁移专用环境配置');
    }
    
    const content = fs.readFileSync(envMigratePath, 'utf8');
    const match = content.match(/DATABASE_URL\s*=\s*"([^"]+)"/);
    if (!match) {
      throw new Error('❌ .env_migrate 中未找到 DATABASE_URL');
    }
    
    return match[1];
  }

  /**
   * 从 .env 获取正常数据库URL
   */
  getNormalDbUrl() {
    const envPath = path.join(process.cwd(), '.env');
    if (!fs.existsSync(envPath)) {
      throw new Error('❌ .env 文件不存在');
    }
    
    const content = fs.readFileSync(envPath, 'utf8');
    const match = content.match(/DATABASE_URL\s*=\s*"([^"]+)"/);
    if (!match) {
      throw new Error('❌ .env 中未找到 DATABASE_URL');
    }
    
    return match[1];
  }

  /**
   * 检查迁移数据库是否存在，如果不存在则创建
   */
  async ensureMigrateDatabase() {
    console.log('🔍 检查迁移数据库...');
    
    try {
      // 解析数据库URL获取数据库名
      const url = new URL(this.migrateDbUrl);
      const dbName = url.pathname.substring(1); // 移除开头的 '/'
      const baseUrl = `${url.protocol}//${url.username}:${url.password}@${url.host}`;
      
      console.log(`📊 迁移数据库: ${dbName}`);
      
      // 创建数据库（如果不存在）
      const createDbCommand = `mysql -h ${url.hostname} -P ${url.port} -u ${url.username} -p${url.password} -e "CREATE DATABASE IF NOT EXISTS ${dbName};"`;
      
      try {
        execSync(createDbCommand, { stdio: 'pipe' });
        console.log('✅ 迁移数据库已准备就绪');
      } catch (error) {
        console.log('⚠️  无法自动创建数据库，请手动创建迁移数据库:', dbName);
      }
      
    } catch (error) {
      console.log('⚠️  数据库URL解析失败，请检查 .env_migrate 配置');
    }
  }

  /**
   * 生成迁移文件
   */
  async generateMigration() {
    console.log('🚀 开始生成迁移文件...');
    console.log(`📝 迁移名称: ${this.migrationName}`);
    console.log(`🔗 使用数据库: ${this.migrateDbUrl}`);

    try {
      // 设置环境变量
      const env = { ...process.env };

      // 读取 .env_migrate 文件并设置环境变量
      const envMigratePath = path.join(process.cwd(), '.env_migrate');
      const envContent = fs.readFileSync(envMigratePath, 'utf8');

      // 解析环境变量
      envContent.split('\n').forEach(line => {
        const trimmed = line.trim();
        if (trimmed && !trimmed.startsWith('#')) {
          const [key, ...valueParts] = trimmed.split('=');
          if (key && valueParts.length > 0) {
            let value = valueParts.join('=').trim();
            // 移除引号
            if ((value.startsWith('"') && value.endsWith('"')) ||
                (value.startsWith("'") && value.endsWith("'"))) {
              value = value.slice(1, -1);
            }
            env[key.trim()] = value;
          }
        }
      });

      const command = `npx prisma migrate dev --name ${this.migrationName}`;
      console.log(`⚡ 执行命令: ${command}`);

      execSync(command, {
        stdio: 'inherit',
        cwd: process.cwd(),
        env: env
      });

      console.log('✅ 迁移文件生成成功');
      return true;
    } catch (error) {
      console.error('❌ 迁移文件生成失败:', error.message);
      return false;
    }
  }

  /**
   * 部署迁移到实际数据库
   */
  async deployMigration() {
    console.log('🚀 开始部署迁移到实际数据库...');
    console.log(`🔗 目标数据库: ${this.normalDbUrl}`);

    try {
      // 设置环境变量
      const env = { ...process.env };

      // 读取 .env.development 文件并设置环境变量
      const envDevPath = path.join(process.cwd(), '.env.development');
      if (fs.existsSync(envDevPath)) {
        const envContent = fs.readFileSync(envDevPath, 'utf8');

        // 解析环境变量
        envContent.split('\n').forEach(line => {
          const trimmed = line.trim();
          if (trimmed && !trimmed.startsWith('#')) {
            const [key, ...valueParts] = trimmed.split('=');
            if (key && valueParts.length > 0) {
              let value = valueParts.join('=').trim();
              // 移除引号
              if ((value.startsWith('"') && value.endsWith('"')) ||
                  (value.startsWith("'") && value.endsWith("'"))) {
                value = value.slice(1, -1);
              }
              env[key.trim()] = value;
            }
          }
        });
      }

      // 同时读取 .env 文件
      const envPath = path.join(process.cwd(), '.env');
      if (fs.existsSync(envPath)) {
        const envContent = fs.readFileSync(envPath, 'utf8');

        // 解析环境变量
        envContent.split('\n').forEach(line => {
          const trimmed = line.trim();
          if (trimmed && !trimmed.startsWith('#')) {
            const [key, ...valueParts] = trimmed.split('=');
            if (key && valueParts.length > 0) {
              let value = valueParts.join('=').trim();
              // 移除引号
              if ((value.startsWith('"') && value.endsWith('"')) ||
                  (value.startsWith("'") && value.endsWith("'"))) {
                value = value.slice(1, -1);
              }
              env[key.trim()] = value;
            }
          }
        });
      }

      const command = `npx prisma migrate deploy`;
      console.log(`⚡ 执行命令: ${command}`);

      execSync(command, {
        stdio: 'inherit',
        cwd: process.cwd(),
        env: env
      });

      console.log('✅ 迁移部署成功');
      return true;
    } catch (error) {
      console.error('❌ 迁移部署失败:', error.message);
      return false;
    }
  }

  /**
   * 执行完整的迁移流程
   */
  async run() {
    console.log('🎯 开始干净迁移流程...');
    console.log('=' .repeat(50));
    
    try {
      // 1. 确保迁移数据库存在
      await this.ensureMigrateDatabase();
      
      // 2. 生成迁移文件
      const generateSuccess = await this.generateMigration();
      if (!generateSuccess) {
        process.exit(1);
      }
      
      console.log('');
      console.log('🔄 迁移文件生成完成，准备部署到实际数据库...');
      console.log('');
      
      // 3. 部署迁移
      const deploySuccess = await this.deployMigration();
      if (!deploySuccess) {
        process.exit(1);
      }
      
      console.log('');
      console.log('🎉 干净迁移流程完成！');
      console.log('✨ 动态表问题已解决，迁移文件已成功应用到实际数据库');
      
    } catch (error) {
      console.error('❌ 迁移流程失败:', error.message);
      process.exit(1);
    }
  }

  /**
   * 显示帮助信息
   */
  static showHelp() {
    console.log(`
🔧 干净迁移工具 - 解决动态表迁移问题

用法:
  npm run prisma:clean-migrate [迁移名称]

示例:
  npm run prisma:clean-migrate add_new_field
  npm run prisma:clean-migrate update_schema

说明:
  此工具使用两阶段迁移流程：
  1. 在干净数据库中生成迁移文件（避免动态表干扰）
  2. 将迁移文件部署到实际数据库

配置:
  - .env_migrate: 迁移专用数据库配置
  - .env: 实际数据库配置
    `);
  }
}

// 主程序
if (require.main === module) {
  if (process.argv.includes('--help') || process.argv.includes('-h')) {
    CleanMigrate.showHelp();
    process.exit(0);
  }
  
  const migrate = new CleanMigrate();
  migrate.run().catch(error => {
    console.error('❌ 程序异常:', error);
    process.exit(1);
  });
}

module.exports = CleanMigrate;
`

**Comment by Richie-Yang** [comment]
> This is a solution for migration offset. I'm not sure if it's useful to you when I have time. For your convenience, the following language is in Chinese. Please translate it yourself

I do have tool to do workaround as well, but i think most people above are requesting this basic feature to be available out of the box for Prisma, not do that by ourself.

**Comment by jude98** [comment]
Hi. 
I was also trying to ignore specified tables/ tables from other schemas while migrations. 
https://www.prisma.io/docs/orm/prisma-schema/data-model/externally-managed-tables

Found out that this preview feature should be working for this use case.

**Comment by matthieuauger** [comment]
I can confirm this preview feature works perfectly for my use-case (ignore DBT-generated tables from Prisma Schema when trying to create new migrations)

**Comment by jcobo-techd** [comment]
Hi.
I agree with the community's latest comments; I think it would be a very interesting feature.

**Comment by matthieuauger** [comment]
@jcobo-techd look at https://github.com/prisma/prisma/issues/8864#issuecomment-3227462462

In my case it solved this issue perfectly

**Comment by jcobo-techd** [comment]
@matthieuauger The problem is that I would still have to go to the ```schema.prisma``` file and define which tables are external to ignore them. And the issue with the project we are developing is that different tables are dynamically created from CSV files. We have one master table, and the rest are dynamic - perhaps not the correct architecture, but that's the way I was told to do it😅

**Comment by CyanoFresh** [comment]
same issue.
I want to ignore all the tables in the schema, except 2-3 of them. These tables are also created by DBT, so syncing 2 different development environments is a huge PITA.
